docs/secrets.md

```diff
@@ -0,0 +1,66 @@
+Notification endpoints will require credentials in order to communicate with the various external services.
+
+We do not want users to include those credentials in plain text into Flux scripts. We need to provide a mechanism for users load secrets from a safe storage mechanism and provide those secrets to the endpoints.
+
+**Backing Store:** : There are many backing store possibilities, which need not be discussed here.  Any store that provides secure access for key/value lookup may be used by providing an implementation of the KeyLookup  interface in the flux repository.  
+
+**Fine Grained Authorization**:  keys will generally not be accessible in the flux language.  A secret-enabled function will accept meta-data about the secret (a key) and use it to look up the stored value in memory. The lookup will error if the key is not found or if the user is not authorized to use that key.  
+
+# **Design**
+
+## Internal Functions
+A function that requires secret lookup must: 
+
+- Be built-in to the language
+- Register an ExternalDependencies struct  (note: it will be straightforward and advantageous to refactor the `InjectFromDependencies` and similar functions into this pattern as well).   
+- Not return secrets under any circumstances (e.g. in error messages) 
+
+The external interface for secret lookup is: 
+```Go
+func LookupRequest(keyObject values.Object) SecretRequest
+func PreAuthLookup(keyObject values.Object) 
+func LookupKey(req SecretRequest) (string, error)
+```
+The keyObject is a values.Object type because: 
+
+- The secret parameter can be passed to the service without being modified
+- The parameter may be validated by the service, minimizing the dependency of the calling function to have prior knowledge of the key lookup protocol. 
+
+The keyObject should minimally have a `secretKey` property: 
+slackID = {secretKey: "FLUX_SLACK_ID"}
+
+but the object may include supplementary information for consumption by the calling function.  They may also be embedded + replaced in another values.Value type (e.g. an encapsulating Object or Array) by the calling function as in:
+
+http.post(header: {"Content Type": "application-json", "ApplicationSecret": {secretKey: "FLUX_HTTP_POST_KEY"}}, url: "http://some/url/", data: json_enc)
+
+## Flux Package
+We will provide a `vault` package that will help the user create an object with the required properties for secret lookup.  
```

Comment:
The package should be called `secrets` or something generic, the fact we are using `vault` should be abstracted away.